InfiniumDiffMetMotR
===================
Version: 1.5.0

Description: This is a R package to analyze transcription factor binding motif enrichment at differentially methylated regions for Infinium Methylation BeadChip (Illumina).  

Last Update: 2019-10-05  

Depends: R (>= 2.10), Biobase (>= 2.5.5)  

Author: Takahiro Suzuki  

Updated by: takahiro.suzuki.aa@riken.jp

## Install
---
## 1. Install of devtools
devtools needs to install packages from github
```r
install.packages("devtools")
```
## 2. Install of InfiniumDiffMetMotR from github
```r
devtools::install_github("takahirosuzuki1980/InfiniumDiffMetMotR")
```
## Examples of usage
***
### ***TF binding Motif overrepresentation analysis***

### 1. Load of InfiniumDiffMetMotR
```r
library("InfiniumDiffMetMotR")
```

### 2. Normalization  
#### - row signal text data (generated by GenomeStudio)
```r
lumiMethyNorm(fileName = "TableControl.txt", inputtype = "signal", sample_names = c("sample1", "sample2"))
```
- ***filename: unnourmalized-no background
correction full infinium methylation array data.***  
The header of the data should includes the following items as the following formats;  
**TargetID    Sample1.Signal_A    Sample1.Signal_B    Sample1.Detection Pval**  
\- TargetID: Illumina probe ID  
\- Sample1.Signal_A: Signal of Unmethylated probe of sample 1  
\- Sample1.Signal_B: Signal of Methylated probe of sample 1  
\- Sample1.Detection Pval: detection P value of sample 1  
- ***inputtype: "signal" for a row signal text data***  
- ***sample_names: a vector of sample names***

#### - idata files
```r
lumiMethyNorm(idatpath=getwd(), inputtype = "idat", sample_names = c("sample1", "sample2"))
```
- ***idatpath: A directory of idat files***  
a set of idat file name should be
**barcode_location_Grn.idat & barcode_location_Red.idat or GEOID_barcode_location_Grn.idat & GEOID_barcode_location_Red.idat (In the case of GEO data)**  
for example  
\- GSM2417753_200598790047_R01C01_Grn.idat  
\- GSM2417753_200598790047_R01C01_Red.idat
- ***inputtype: "idat" for idat files***  
- ***sample_names: a vector of sample names. If idat files are downloaded from GEO, sample_names should be "FALSE". Then, sample names are automatically downloaded and converted to GEO sample title baed on the file name.***


*lumiMethyNorm generates **Process_Result folder**: reports of the normalization, **processed_Mval.txt**: a matrix of M-value, and **sel_processed_Mval.txt**: a matrix of M-value which does not inculude low detection p-value probes. (default cut off is 0.01)*  


  
### 3. motif database construction  
#### - Example 1: JASPER_CORE, Hsapiencs and Mmusclus
```r
library("MotifDb")
targetDB <- "JASPAR_CORE"
targetORG <- c("Hsapiens", "Mmusculus")
motifDB <- query(MotifDb, targetDB)        #extraction of motif list of "JASPER_CORE"
motifDB <- c(query(motifDB,targetORG[1]),query(motifDB,targetORG[2]))        #extraction of motifs of "Hsapiens" and "Mmusclus"
```
If you want analyze a specific motif select a motif. (ex. SPI1)
```r
targetTF <- "SPI1"
motifDB <- query(motifDB,targetTF)       #Extraction of motifs for target TF(s)
```
Finally, convert the motif list to list format.
```r
motifDBList <- as.list(motifDB)
```
#### - Example 2: IMAGE motif database
```r
motifDBList <- IMAGE_PWMlist
```

Motif list should be a list of PWMs of following format;  
  
   1     2    3    4    5    6    7    8    9  
A    0    0    0    0.1189189    0.1027027    0.2972973    0.28648649    0.10270270    0.04864865  
C    0    1    1    0.3837838    0.3081081    0.2378378    0.16216216    0.08648649    0.42162162  
G    1    0    0    0.2486486    0.3297297    0.3621622    0.49189189    0.74054054    0.42702703  
T    0    0    0    0.2486486    0.2594595    0.1027027    0.05945946    0.07027027    0.10270270  
  
 
### 4. Screening of enriched motifs  
```r
MotScr(infile="sel_processed_Mval.txt", motifDBList = motifDBList, cutoff = 2, p.cutoff = 0.001, outname="screening_result", ControlColnum=c(1,2), TreatmentColnum=c(3,4), MethylDemethyl="Demethyl", sampling=FALSE, version = "850")
```
- ***infile: M-value matrix of infinium methylation array***  
- ***motifDBList: list format of PWM***  
- ***cutoff: cutoff velue of delta M***  
- ***outname: name of output file***  
- ***ControlColnum: A column of control data, such as unddiferentiated***  
- ***TreatmentColnum: A column of treatment data such as differentiated***  
- ***MethylDemethyl: "Methyl" (or "M") or "dimethyl (or"D").analysis target of differentially methylated probes***
- ***sampling: number of randomly sampled differentiall methylated probes. "FALSE" analyses all differentially methylated probes***  
- ***version: "450" or "EPIC" (or "850"). version of methylation array***  


*If you perform a comparison of multiple samples, identification of differentially methylated probes uses both Welch's t-test and M-value difference. For single sample comparison, it uses only M-value difference (delta M).*  

If you want to extract the motif of interest (e.g. SPI1):
```r
motif_names <- "SPI1"
motif <- motifDBList[grep(motif_names,names(motifDBList))]
MotScr(infile="sel_processed_Mval.txt", motifDBList = motif, cutoff = 2, p.cutoff = 0.001, outname="screening_result", ControlColnum=c(1,2), TreatmentColnum=c(3,4), MethylDemethyl="Demethyl", sampling=FALSE, version = "850")
```


 ### 5. output files
 - [outname_DMP_position.txt  
 DMP position
 - [outname]_mot_analysis_result.txt  
summary table
 - [outname]_plot.pdf  
 All histograms, enrichment score plots, fold-change plots, and p-value plots
 - [outname]_result.RData  
R data file
 - [outname]_sig_plots directory/[motif_name].pdf  
A directory containing enrichment score plots of significantly enriched motifs

### ***Scatter plot***
### 1. Reading of M-value data  
```r
infile <- "sel_processed_Mval.txt"
selDataMatrix <- read.table (infile)
```
### 2. Labels of the plot  
```r
main <- paste(colnames(selDataMatrix)[ControlColnum]," vs. ", colnames(selDataMatrix)[TreatmentColnum], sep="")
xlab <- paste("Control: ", colnames(selDataMatrix)[ControlColnum], sep="")
ylab <- paste("Treatment: ", colnames(selDataMatrix)[TreatmentColnum], sep="")
pdf ("Scatter_Plot.pdf")
```
### 3. Plot  
```r
scatterPlot(treatment_data=selDataMatrix[,8], control_data=selDataMatrix[,1], main=main, xlab=xlab, ylab=ylab, cutoff=2)
dev.off()
```